host_core/lib/host_core/providers/provider_module.ex

```diff
@@ -146,21 +200,50 @@ defmodule HostCore.Providers.ProviderModule do
 
   @impl true
   def handle_call(:halt_cleanup, _from, state) do
-    # Elixir cleans up ports, but it doesn't always clean up the OS process it created
-    # for that port. TODO - find a clean, reliable way of killing these processes.
-    if state.os_pid != nil do
-      System.cmd("kill", ["-9", "#{state.os_pid}"])
-    end
+    Logger.debug("Provider termination requested manually")
+
+    Tracer.with_span "Terminate provider instance", kind: :server do
+      case HostCore.Nats.safe_req(
+             HostCore.Nats.rpc_connection(state.lattice_prefix),
+             "wasmbus.rpc.#{state.lattice_prefix}.#{state.public_key}.#{state.link_name}.shutdown",
+             Jason.encode!(%{host_id: state.host_id}),
+             receive_timeout: 2000
+           ) do
+        {:ok, _msg} ->
+          Logger.debug("Provider acknowledged shutdown request")
+
+        {:error, :no_responders} ->
+          Logger.error("No responders to RPC request to terminate provider")
+
+        {:error, :timeout} ->
+          Logger.error("No capability providers responded to RPC shutdown request")
+      end
 
-    publish_provider_stopped(
-      state.public_key,
-      state.link_name,
-      state.instance_id,
-      state.contract_id,
-      "normal"
-    )
+      # Pause for n milliseconds between shutdown request and forceful termination
+      # Process.sleep(state.shutdown_delay)
+
+      # Elixir cleans up ports, but it doesn't always clean up the OS process it created
+      # for that port. TODO - find a clean, reliable way of killing these processes.
+      if state.os_pid != nil do
+        # fun fact - if we don't do this in a spawned task, we never move execution
+        # to after the if statement. HUZZAH
+        Task.Supervisor.start_child(ControlInterfaceTaskSupervisor, fn ->
+          System.cmd("kill", ["-9", "#{state.os_pid}"])
```

Comment:
Does this work on all platforms?

Reply:
yes, always has. The `System.cmd` call isn't new code.